Distinct Strings and SemiExplicitConverter

uwieder opened the pull request:
I created a DistinctStrings class and a DistinctStringsSketch. The DistinctStrings class can create a SemiExplicitConverter which is just an explicit converter with a default value.
In TestUtil I added a static createTable function that creates a table with 3 columns of different types and different distributions. I test this in DistinctStringsSketchTest.

mbudiu-vmw commented on hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/DistinctStrings.java:
```diff
@@ -0,0 +1,59 @@
+        int newMaxSize = Integer.max(this.maxSize, otherSet.maxSize);
+        result.addAll(this.mySet);
+        result.addAll(otherSet.mySet);
+        while(result.size() > newMaxSize) {
```
Space missing.
Maybe we need some indication that we have dropped data - a boolean flag?

mbudiu-vmw commented on hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/DistinctStrings.java:
```diff
@@ -0,0 +1,59 @@
+    public SemiExplicitConverter  getStringConverter() {
+        SemiExplicitConverter converter = new SemiExplicitConverter(this.mySet.size() + 1);
+        int i = 1;
+        while(!this.mySet.isEmpty()) {
```
Missing a space.
It may be more efficient to iterate over the set; removing from the set may be unnecessarily expensive.

mbudiu-vmw commented on hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/DistinctStrings.java:
```diff
@@ -0,0 +1,59 @@
+
+    public SemiExplicitConverter  getStringConverter() {
+        SemiExplicitConverter converter = new SemiExplicitConverter(this.mySet.size() + 1);
+        int i = 1;
```
Why start at 1?

mbudiu-vmw commented on hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/DistinctStrings.java:
```diff
@@ -0,0 +1,59 @@
+    private final int maxSize;
+    private final TreeSet<String> mySet;
+
+    public DistinctStrings(final int maxSize) {
```
How about if maxSize is 0 there is no limit?
Please document this behavior.

mbudiu-vmw commented on hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/DistinctStrings.java:
```diff
@@ -0,0 +1,59 @@
+        this.mySet = mySet;
+    }
+
+    public void getStrings(final IColumn column, final IMembershipSet membershipSet) {
```
I would call this method addStrings...
Do you want to add a check that the column is a string column?

mbudiu-vmw commented on hieroplatform/src/main/java/org/hiero/sketch/table/BaseExplicitConverter.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.util.HashMap;
+
+public abstract class BaseExplicitConverter implements IStringConverter {
```
Please move the javadoc comment from ExplicitStringConverter here.

mbudiu-vmw commented on hieroplatform/src/main/java/org/hiero/sketch/table/SemiExplicitConverter.java:
```diff
@@ -0,0 +1,18 @@
+package org.hiero.sketch.table;
+
+public class SemiExplicitConverter extends BaseExplicitConverter {
+    final double defaultValue;
```
How about we make defaultValue be the constant -1?
One less thing to keep track of.

uwieder commented:
I committed the changes to my branch converters. I fixed a bug in the getConverter method. For some reason previous version destroyed the DistinctStrings object while creating the converter. I fixed that. However, after merging my tests *FAIL* on a RemotingTest.
